stm32: pm: Enable to driver JTAG port pns to analog when not required

erwango opened the pull request:
Configuring JTAG port pins in analog when not required (build time) can save few uA.

Requires https://github.com/zephyrproject-rtos/zephyr/pull/64109

Note:
This operation can also be done using pinctrl API, which would allow dynamic handling of the pins (JTAG function are alternate functions which can't be configured using GPIO API)
Main reason to push this PR using GPIO API is that this is my first shot at this. If everyone agrees this is good enough, this is fine with me

erwango committed:
boards: nucleo_wba52cg: Enable flash and debug using OpenOCD

OpenOCD can now be used to flash and debug nucleo_wba52cg.
However it required use of STMicroelectronics OpenOCD fork.

Add instructions on how to use it.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

erwango committed:
soc: stm32: PM: Disable jtag port pins if no debug

At chip startup, jtag pins are configured by default to enable
debug.
This configuration adds consumption and when using PM profile,
we can save ~40uA by resetting this configuration and setting pins
to analog mode.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

erwango committed:
dts: stm32u5: Add SW/JTAG debug port node

Provide jtag port pins description, so they can be used to be set in
analog mode when not required to save power (around 40uA saved in total).

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

gmarull commented on dts/arm/st/wba/stm32wba.dtsi:
```diff
@@ -98,6 +98,12 @@
 		};
 	};
 
+	jtag_port: jtag_port {
+		compatible = "gpio-jtag";
+		jtag-gpios = <&gpioa 13 0>, <&gpioa 14 0>, <&gpioa 15 0>,
+			     <&gpiob 3 0> ,<&gpiob 4 0>;
+	};
```
I'm not sure the "gpio-jtag" is a well-defined binding. The fact you're using GPIO here is because you want to use the Zephyr GPIO API, not because this necessarily describes the hardware well. In fact, I find it a bit confusing as I have no clue by reading devicetree what all those pins mean. Which one is TDI, TMS...?

I'd suggest exploring a solution using pinctrl, would make things easier as well (technically you do not need a driver to use it, just use `PINCTRL_DT_DEFINE(DT_NODELABEL(jtag_port))`.

erwango replied:
> I'd suggest exploring a solution using pinctrl, would make things easier as well

Indeed. See my note in the PR description. Thanks for confirming my doubt.

> In fact, I find it a bit confusing as I have no clue by reading devicetree what all those pins mean. Which one is TDI, TMS...?

Tbh, if they are not used otherwise, I'm not sure this is useful. Note that even with pinctrl, it is difficult to extract one single pin. So being able to identify them is nice but, I don't see how it could be used.

> (technically you do not need a driver to use it, just use PINCTRL_DT_DEFINE(DT_NODELABEL(jtag_port)).

Yes, but I was wondering where to actually use it. Maybe I can just reuse the same bootstrap that I'm using currently with GPIO in the end.

gmarull replied:
I guess you could have something like:


```
jtag: connector {
    compatible = "jtag-connector";
    pinctrl-0 = <&jtag_tms_pa13...>;
    pinctrl-1 = <&analog_pa13...>;
    pinctrl-names = "default", "sleep";
};
```

erwango replied:
I've started having a look. Regarding AF, no issue, script handles it quite easily.
Though, each pin has its own specific configuration (speed, pupdr), which is not documented in database.
So, we have 2 options:
- Fill speed and state based on ref man information specifically for each pin (hopefully configuration doesn't depend on series). Gently tedious but doable thanks to the nice script we have but I've heard some are not quite happy with filling pin properties in -pcintrl.dtsi files ;)
- Don't fill the the pin properties, assuming they won't be used (ie, no dynamic switching from analog back to functional debugging state > I wonder about the interest of this use case anyway).

gmarull replied:
> I've started having a look. Regarding AF, no issue, script handles it quite easily. Though, each pin has its own specific configuration (speed, pupdr), which is not documented in database. So, we have 2 options:
> 
> * Fill speed and state based on ref man information specifically for each pin (hopefully configuration doesn't depend on series). Gently tedious but doable thanks to the nice script we have but I've heard some are not quite happy with filling pin properties in -pcintrl.dtsi files ;)

That script, in retrospective, was a mistake that propagated from the very early board-based pinctrl model. The solution implemented for GD32 is far better and more flexible (it's a copy of Linux STM32 model), the script simply generates all AFs, no need to worry if new peripherals are added, all definitions will already be there. Properties are up to the user to figure out, same as you'd do developing with the HAL in bare-metal. I'd suggest reviewing the ST model to at least drop the pin properties from the generated files if this is a problem.

erwango replied:
> That script, in retrospective, [...]

I didn't wanted to re-open the discussion on this. On my side I'm still not convinced about the scalability of this model. 

> it's a copy of Linux STM32 model

Discussing with Linux STM32 folks, they agree current Zephyr solution is much more suited to our "lot of SoCs, lot of pin combinations" model, and some would even prefer to use this over Linux model (not saying it will happen any day).

> "HAL in bare-metal" 

HAL is not a model of scalability. Besides, from the reports I get from users, I'm not convinced either that STM32 Linux model would be more easily handled than current definitions. There were some issues reported but none that I can remember about the fact of having pin properties being defined in -pinctrl.dtsi files

I can understand that this model is not conceptually perfect, but as long as it will be easy to maintain and there aren't significant issues directly linked to it, I don't see reasons to change. If it happen one day that, for instance, dts pin config is generated from a tool making usage and maintenance eased by a different model, I'll be open to discussion.

Coming back to current issue, I don't see how STM32 Linux model would be of any help.
For now I'll go with solution 2, adding a warning on the fact that it is not designed to swap pins dynamically (but can be updated in board if required). This can be updated later if required.

(Damn, for someone not willing to enter the discussion, I've failed :-) )

gmarull replied:
I guess Linux folks just miss the pre-defined pin combinations, getting AFs from datasheet can be painful. The properties are usually trivial to guess (e.g. pull-ups, open-drain, drive, etc.). So the GD32 is the Linux STM32 model, but better in the sense that you don't need to look for AFs.

Anyway, I won't insist on this, in the end it's there for historical reasons. But no other vendor has been allowed to do the same because it often leads to issues or misconfigurations (e.g. ST still hasn't fixed the I2C case).

gmarull commented on soc/arm/st_stm32/common/pm_debug_jtag.c:
```diff
@@ -0,0 +1,33 @@
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/init.h>
+
+#if !defined(CONFIG_DEBUG) && defined(CONFIG_PM)
+
+struct gpio_dt_spec jtag_port[] = {
+	DT_FOREACH_PROP_ELEM_SEP(DT_NODELABEL(jtag_port), jtag_gpios, GPIO_DT_SPEC_GET_BY_IDX, (,))
+};
+
+static int jtag_to_analog(void)
+{
+
+	/* Set JTAG port pins to analog mode */
+
+	for (int i = 0; i < ARRAY_SIZE(jtag_port); i++) {
+		if (!gpio_is_ready_dt(&jtag_port[i])) {
+			return -ENODEV;
+		}
+		(void)gpio_pin_configure_dt(&jtag_port[i], GPIO_DISCONNECTED);
+	}
+
+	return 0;
+}
+
+SYS_INIT(jtag_to_analog, POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE);
+
+#endif
```
-1, impose whole file Kconfig conditions at CMake level

gmarull commented on soc/arm/st_stm32/common/pm_debug_jtag.c:
```diff
@@ -0,0 +1,33 @@
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/init.h>
+
+#if !defined(CONFIG_DEBUG) && defined(CONFIG_PM)
```
is `CONFIG_PM` a strict condition here?

erwango replied:
If `CONFIG_PM` is not a condition, it would mean that you can't use gdb unless CONFIG_DEBUG is set, which is not a condition today.
I already foresee that I will get tens of questions around this..
Using `CONFIG_PM`, I don't break anything as gdb can't be used today with CONFIG_PM unless CONFIG_DEBUG is set.

erwango committed:
west.yml: hal_stm32: Use SW/JTAG signals description

SW/JTAG signals description are useful when we don't need them.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

gmarull commented on soc/arm/st_stm32/common/pm_debug_swj.c:
```diff
@@ -0,0 +1,36 @@
+	int err;
+
+	/* Set Serial Wire / JTAG port pins to analog mode */
+	err = pinctrl_apply_state(swj_pcfg, PINCTRL_STATE_SLEEP);
```
note: I remember now that sleep state requires CONFIG_PM_DEVICE=y, CONFIG_PM alone will not make that state available.

erwango replied:
I wonder how I got it working then. Wait. I just tested that debug wasn't functional and not measured consumption.. Strange anyway. I'll have a new check.
Then, do you agree to make it available with CONFIG_PM=y only then ? My case, is not limited to PM_DEVICE.

erwango committed:
include: pinctrl.h: Make PINCTRL_SKIP_SLEEP available with CONFIG_PM

In some cases, PINCTRL_STATE_SLEEP may be required even when
CONFIG_PM_DEVICE is not defined.
One example is the possibility/need to put JTAG pins to analog when
CONIG_PM=y and CONFIG_DEBUG=n.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

erwango commented on dts/arm/st/wba/stm32wba.dtsi:
```diff
@@ -98,6 +98,12 @@
 		};
 	};
 
+	jtag_port: jtag_port {
+		compatible = "gpio-jtag";
+		jtag-gpios = <&gpioa 13 0>, <&gpioa 14 0>, <&gpioa 15 0>,
+			     <&gpiob 3 0> ,<&gpiob 4 0>;
```
Leftover from initial proposal. To be removed

gautierg-st commented on west.yml:
```diff
@@ -229,7 +229,7 @@ manifest:
         - hal
     - name: hal_stm32
-      revision: 22925907a6faeb601fc9a0d8cbb65c4b26d38043
+      revision: pull/180/head
```
You can update the revision now.

gautierg-st commented on dts/arm/st/wba/stm32wba.dtsi:
```diff
@@ -462,6 +462,16 @@
+	swj_port: swj_port {
+		compatible = "swj-connector";
+		pinctrl-0 = <&debug_jtms_swdio_pa13 &debug_jtck_swclk_pa14
+			     &debug_jtdi_pa15 &debug_jtdo_swo_pb3
+			     &debug_jtrst_pb4>;
+		pinctrl-1 = <&analog_pa13 &analog_pa14 &analog_pa15
+			     &analog_pb3 &analog_pb4>;
```
I'm not sure the dtsi is the best place to put that since the pins can be used for something else than debug.
For example, in the nucleo_wba55cg.dts, PA15, PB3 and PB4 are used for SPI1. Depending on the order of initialization, I think it could result in drivers not functioning.

erwango replied:
Right, so we should make sure this is done before configuring any other device and change priority
